File: ETLsWithR/Projects/Catalogs/Libraries/SitesCatalog.R
#*******************************************************************************
#* SitesCatalog.R
#* Sites Catalog
#* Carlos Kassab
#* 2019-Jan-28
#*******************************************************************************


#*******************************************************************************
# This function may be called to execute any code before the sql query for data
# extraction.
#*******************************************************************************
BeforeDataExtract = function( srcConn, trgtConn, procRunNum, projectID, procNum ) {
  
  tryCatch( {
    print( "Do something and return a value." )
    return( c(1,2,3,4) )
  }, error = function( ex ) {
    print( paste( "Cannot run process:", ex ) )
    return( FALSE )
  } )   
  
}



#*******************************************************************************
# This function is to extract data from source database, sites catalog table by 
# executing the select query in queries script.
#*******************************************************************************

DataExtract = function( srcConn, trgtConn, procRunNum, projectID, procNum, getDataQuery ) {
  
  tryCatch( {
    WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 210, 1, 0, '' ) # Start.
    sourceData = dbGetQuery( srcConn, getDataQuery )
    names(sourceData)[names(sourceData) == 'ID'] <- 'SiteID' 
    names(sourceData)[names(sourceData) == 'SITE'] <- 'Site'
    names(sourceData)[names(sourceData) == 'DESCRIPTION'] <- 'Description'
    WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 210, 2, 0, '' ) # End
    return( sourceData )
  }, error = function( ex ) {
    WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 210, 3, 3003, paste( "Error reading sites catalog:", ex ) )
    return( FALSE )
  } )   
  
}



#*******************************************************************************
# Process data after data was read from source table.
#*******************************************************************************

AfterDataExtract = function( dataExtracted, trgtConn, procRunNum, projectID
                            , procId, deleteQuery ) {
  
  tryCatch( {
    # Truncate Table on Target DB.
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 215, 1, 0, '' )
    truncateResult = dbSendQuery( trgtConn, deleteQuery )
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 215, 2, 0, '' )
  }, error = function( ex ) {
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 215, 3, 3003, paste( "Error truncating sites table:", ex ) )
    return( FALSE )
  } )   

  tryCatch( {
    # Insert New Data Into Target Table.
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 220, 1, 0, '' )
    insertResult = dbWriteTable( trgtConn, "CAT_Sites", dataExtracted, append = T, row.names = FALSE )
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 220, 2, 0, '' )
    return( TRUE )
  }, error = function( ex ) {
    WriteETLMSLog( trgtConn, procRunNum, projectID, procId, 220, 3, 3003, paste( "Error inserting into sites table:", ex ) )
    return( FALSE )
  } )   
  
}



#*******************************************************************************
# After Loading This File, Call This function to process data.
#*******************************************************************************

MainProcess = function( srcConn, trgtConn, procRunNum, projectID, procId ){
  tryCatch( {
    source( "../Queries/SitesQueries.R" )

    dataExtracted = DataExtract( srcConn, trgtConn, procRunNum, projectID, procId, selectQuery )
    
    if( typeof( dataExtracted ) == "logical" ) return( FALSE )
    
    if( !AfterDataExtract( dataExtracted, trgtConn, procRunNum, projectID, procId, truncateQuery ) ) return( FALSE )

    return( TRUE )
  }, error = function( ex ) {
    print( paste( "Cannot read sites query data:", ex ) )
    return( FALSE )
  } )
  
}

File: ETLsWithScala/Projects/Catalogs/.scala-build/Catalogs_c34520776b-e80abaebed/src_generated/main/SitesCatalog.scala


final class SitesCatalog$_ {
def args = SitesCatalog_sc.args$
def scriptPath = """SitesCatalog.sc"""
/*<script>*/


// //> using file ../../../GeneralLibraries/Utils.sc
//import Utils_sc.script

import java.sql._

/*******************************************************************************
 This method may be called to execute any logic before the sql query for data
 extraction.
********************************************************************************/
override def BeforeDataExtract(srcConn: Connection, trgtConn: Connection, procRunNum: String, projectID: Int, procNum: Int): Any = {
  try {
    println( "Do something and return a value." )
    return 3333
  } catch {
    case error: Exception => {
      println(s"Cannot execute BeforeDataExtract:${error.getMessage()}")
      return -1
    }
  }  
}


/*******************************************************************************
 This method is to extract data from source database, sites catalog table by 
 executing the select query in queries script.
********************************************************************************/
override def DataExtract(srcConn: Connection, trgtConn: Connection, procRunNum: String, projectID: Int, procNum: Int, getDataQuery: String, logFileName: String ): ResultSet = {
  
  val sourceData = try { // Using EOP(Expression Oriented Programming) style
    println( "Extract Data from Source DataBase." )
    //WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 200, 1, 0, "", logFileName ) // Start.
    //WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 210, 1, 0, "", logFileName ) // Start.
    //Utils.WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 220, 1, 0, "", logFileName ) // Running.
    val stmt = srcConn.createStatement()
    stmt.executeQuery(getDataQuery)
    stmt.getResultSet
  } catch {
    case error: Exception => {
      println(s"Cannot execute DataExtract:${error.getMessage()}")
      return null
    }
  }
  return sourceData
}





override def MainProcess( srcConn: Connection, trgtConn: Connection, procRunNum: String, projectID: Int, procNum: Int, logFileName: String, writeETLLog: (Int, Int, Int, String) => Unit ): Boolean = {
  try {
    //> using file Queries/SitesQueries.sc
    //println( "Extract Data from Source DataBase and Load it to Target DataBase." )
    //writeETLLog.asInstanceOf[( Connection, String, Int, Int, Int, Int, Int, String, String ) => Unit]
    //  .apply( trgtConn, procRunNum, projectID, procNum, 210, 1, 0, "", logFileName ) // Start.
    writeETLLog(210, 1, 0, "")
    //Utils.WriteETLMSLog( trgtConn, procRunNum, projectID, procNum, 0, 1, 0, "", logFileName ) // Start.
    // Call the MainProcess method from SitesCatalog library
    return true
  } catch {
    case error: Exception => {
      println(s"Cannot execute MainProcess:${error.getMessage()}")
      return false
    }
  }  
}








/*</script>*/ /*<generated>*//*</generated>*/
}

object SitesCatalog_sc {
  private var args$opt0 = Option.empty[Array[String]]
  def args$set(args: Array[String]): Unit = {
    args$opt0 = Some(args)
  }
  def args$opt: Option[Array[String]] = args$opt0
  def args$: Array[String] = args$opt.getOrElse {
    sys.error("No arguments passed to this script")
  }

  lazy val script = new SitesCatalog$_

  def main(args: Array[String]): Unit = {
    args$set(args)
    val _ = script.hashCode() // hashCode to clear scalac warning about pure expression in statement position
  }
}

export SitesCatalog_sc.script as `SitesCatalog`

